mea/src/mpsc/mod.rs

```diff
@@ -0,0 +1,310 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A multi-producer, single-consumer queue for sending values between asynchronous tasks.
+
+use std::fmt;
+use std::future::poll_fn;
+use std::ptr::null_mut;
+use std::sync::atomic::AtomicPtr;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::task::Context;
+use std::task::Poll;
+use std::task::Waker;
+
+#[cfg(test)]
+mod tests;
+
+/// Creates an unbounded mpsc channel for communicating between asynchronous
+/// tasks without backpressure.
+///
+/// A `send` on this channel will always succeed as long as the receiver is alive.
+/// If the receiver falls behind, messages will be arbitrarily buffered.
+///
+/// Note that the amount of available system memory is an implicit bound to
+/// the channel. Using an `unbounded` channel has the ability of causing the
+/// process to run out of memory. In this case, the process will be aborted.
+pub fn unbounded<T>() -> (UnboundedSender<T>, UnboundedReceiver<T>) {
+    let state = Arc::new(UnboundedState {
+        senders: AtomicUsize::new(1),
+        rx_task: AtomicPtr::new(null_mut()),
+    });
+    let (sender, receiver) = std::sync::mpsc::channel();
+    let sender = UnboundedSender {
+        state: state.clone(),
+        sender: Some(sender),
+    };
+    let receiver = UnboundedReceiver {
+        state: state.clone(),
+        receiver,
+    };
+    (sender, receiver)
+}
+
+struct UnboundedState {
+    senders: AtomicUsize,
+    rx_task: AtomicPtr<Waker>,
+}
+
+/// Send values to the associated [`UnboundedReceiver`].
+///
+/// Instances are created by the [`unbounded`] function.
+pub struct UnboundedSender<T> {
+    state: Arc<UnboundedState>,
+    sender: Option<std::sync::mpsc::Sender<T>>,
+}
+
+impl<T> Clone for UnboundedSender<T> {
+    fn clone(&self) -> Self {
+        self.state.senders.fetch_add(1, Ordering::Release);
+        UnboundedSender {
+            state: self.state.clone(),
+            sender: self.sender.clone(),
+        }
+    }
+}
+
+impl<T> fmt::Debug for UnboundedSender<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("UnboundedSender").finish_non_exhaustive()
+    }
+}
+
+impl<T> Drop for UnboundedSender<T> {
+    fn drop(&mut self) {
+        drop(self.sender.take()); // drop the sender; this closes the channel if this is the last sender
+
+        match self.state.senders.fetch_sub(1, Ordering::AcqRel) {
+            1 => {
+                // If this is the last sender, we need to wake up the receiver
+                // so it can observe the disconnected state.
+                let waker = self.state.rx_task.swap(null_mut(), Ordering::AcqRel);
+                if !waker.is_null() {
+                    unsafe { (*waker).wake_by_ref() }
```

Comment:
This will leak both the `Waker` and the `Box` it was stored in.

Reply:
Thanks for pointing this out!

These two should be able to fix by replacing with `unsafe { Box::from_raw(waker).wake() }`.